rfd/0057-automatic-aws-server-discovery.md

```diff
@@ -235,6 +235,153 @@ discovery_service:
+    ssm:
+      # default to this as a result of agentless: true
+      document_name: "TeleportAgentlessDiscoveryInstaller"
+```
+
+
+An agentless specific SSM document will be required. The `teleport discovery bootstrap`
+command will need to be updated to create SSM documents appropriate for agentless discovery.
+
+Example SSM document:
+```yaml
+# name: TeleportAgentlessDiscoveryInstaller
+---
+schemaVersion: '2.2'
+description: aws:runShellScript
+parameters:
+  caKey:
+    types: String
+    description: "(Required) The Teleport CA to sshd will trust."
+  hostCert:
+    types: String
+    description: "(Required) The Teleport host cert to use."
+  sshdConfigPath:
+    types: String
+    description: "(Required) The path to the sshd config file."
+mainSteps:
+- action: aws:downloadContent
+  name: downloadContent
+  inputs:
+    sourceType: "HTTP"
+    destinationPath: "/tmp/installTeleport.sh"
+    sourceInfo:
+      url: "https://teleportcluster.xyz/webapi/scripts/default-agentless-installer"
+- action: aws:runShellScript
+  name: runShellScript
+  inputs:
+    timeoutSeconds: '300'
+    runCommand:
+      - export CERT_SECRETS='{{ secretName }}'
+      - export SSHD_CONFIG='{{ sshdConfigPath }}'
+      - /bin/sh /tmp/installTeleport.sh
+```
+
+Agentless mode will serve a different install script resource named
+`default-agentless-installer`. Which will be used to update and restart the sshd
+configuration.
+
+Possible agentless installer script:
+
+```bash
+(
+  flock -n 9 || exit 1
+
+  if grep -q 'TrustedUserCAKeys /etc/ssh/teleport_user_ca.pub' "$SSHD_CONFIG"; then
+	exit 0
+  fi
+
+  if [ "$distro_id" = "debian" ] || [ "$distro_id" = "ubuntu" ]; then
+	sudo apt-get install -y awscli jq
+  elif [ "$distro_id" = "amzn" ] || [ "$distro_id" = "rhel" ]; then
+    sudo yum install -y awscli jq
+  fi
+
+  CA_KEY=$(aws secretsmanager get-secret-value --output=json --secret-id="$CERT_SECRETS" | jq -r '.["SecretString"] | fromjson."ca_key"')
+  HOST_KEY=$(aws secretsmanager get-secret-value --output=json --secret-id="$CERT_SECRETS" | jq -r '.["SecretString"] | fromjson."host_key"')
+  HOST_CERT=$(aws secretsmanager get-secret-value --output=json --secret-id="$CERT_SECRETS" | jq -r '.["SecretString"] | fromjson."host_cert"')
+
+  echo "$CA_KEY" > /etc/ssh/teleport_user_ca.pub
+  echo 'TrustedUserCAKeys /etc/ssh/teleport_user_ca.pub' >> $SSHD_CONFIG
+
+  echo "$HOST_KEY" > /etc/ssh/teleport
+  chmod 0600 /etc/ssh/teleport
+  echo "$HOST_CERT" > /etc/ssh/teleport-cert.pub
+  chmod 0600 /etc/ssh/teleport-cert.pub
+
+  echo 'HostKey /etc/ssh/teleport' >> $SSHD_CONFIG
+  echo 'HostCertificate /etc/ssh/teleport-cert.pub' >> $SSHD_CONFIG
+
+  if ! sshd -t; then
+	echo "sshd_config is in a bad state, exiting without reloading"
+	exit 1
+  fi
+  systemctl restart sshd
+
+) 9>/var/lock/teleport_install.lock
+```
+
+The discovery agent will need CRUD permissions on `host_cert` resources as well as
+permission to create Node resources.
+
```

Comment:
@lxea Please add a section about CA rotation. We can keep it out of initial scope of implementation but we should address it in the RFD.